chainer/functions/math/sparse_matmul.py

```diff
@@ -0,0 +1,432 @@
+    if dtype is None:
+        dtype = numpy.result_type(sp_data.dtype, dn.dtype)
+
+    A_data = sp_data
+    if transa:
+        A_row = sp_col
+        A_col = sp_row
+        A_shape = (sp_shape[1], sp_shape[0])
+    else:
+        A_row = sp_row
+        A_col = sp_col
+        A_shape = sp_shape
+    if transb:
+        B = dn.swapaxes(-1, -2)
+    else:
+        B = dn
+
+    xp = cuda.get_array_module(A_data, B)
+    if xp is numpy:
+        C = _sparse_matmul_cpu(A_data, A_row, A_col, A_shape, B, dtype)
+    else:
+        C = _sparse_matmul_gpu(A_data, A_row, A_col, A_shape, B, dtype)
+
+    if transc:
+        C = C.swapaxes(-1, -2)
+    return C
+
+
+def _sparse_matmul_cpu(A_data, A_row, A_col, A_shape, B, dtype):
+    # A_shape: (_m, _k)
+    # B.shape: ((nb,) _k, _n)
+    # A_data/row/col.shape: ((nb,) ldnz)
+    if not _scipy_available:
+        msg = "SciPy seems to be unavailable on your system. A CPU" \
+              " implementation of sparse_matmul uses SciPy, so you" \
+              " cannot use sparse_matmul on the CPU."
+        raise RuntimeError(msg)
+
+    _m, _k = A_shape
+    _n = B.shape[-1]
+    if B.ndim == 2:
+        sp_A = sparse.coo_matrix((A_data, (A_row, A_col)), shape=(_m, _k))
+        C = sp_A.dot(B).astype(dtype, copy=False)
+    else:
+        nb = B.shape[0]
+        C = numpy.empty((nb, _m, _n), dtype=dtype)
+        for i in range(nb):
+            nnz = len(numpy.where(A_row[i] >= 0)[0])
+            sp_A = sparse.coo_matrix((A_data[i, :nnz],
+                                      (A_row[i, :nnz], A_col[i, :nnz])),
+                                     shape=(_m, _k))
+            C[i] = sp_A.dot(B[i]).astype(dtype, copy=False)
+
+    return C
+
+
+def _sparse_matmul_gpu(A_data, A_row, A_col, A_shape, B, dtype):
+    cupy_dtype = dtype
+    if cupy_dtype == numpy.float16:
+        cupy_dtype = numpy.float32
+        # fp32 is used in cupy kernel because fp16 atomicAdd is not supported
+
+    # A_shape: (_m, _k)
+    # B.shape: ((nb,) _k, _n)
+    # A_data/row/col.shape: ((nb,) ldnz)
+    _m, _k = A_shape
+    _n = B.shape[-1]
+    ldnz = A_data.shape[-1]
+    if B.ndim == 2:
+        nb = 1
+        C = cuda.cupy.zeros((_m, _n), dtype=cupy_dtype)
+    else:
+        nb = B.shape[0]
+        C = cuda.cupy.zeros((nb, _m, _n), dtype=cupy_dtype)
+
+    nthreads = nb * ldnz * _n
+    _cupy_sparse_matmul()(nb, _m, _n, _k, ldnz, A_data, A_row, A_col,
+                          B, C, size=nthreads)
+
+    return C.astype(dtype, copy=False)
+
+
+def _cupy_sparse_matmul():
+    return cuda.cupy.ElementwiseKernel(
+        'int32 nb, int32 _m, int32 _n, int32 _k, int32 nnz, \
+         raw A A_data, raw T A_row, raw T A_col, \
+         raw B _B, raw C _C',
+        '',
+        '''
+        int i_n = (i % _n);
+        int i_A = (i / _n);
+        int i_b = (i / _n) / nnz;
+        if (i_b >= nb) {
+            continue;
+        }
+        int i_k = A_col[i_A];
+        if (i_k < 0) {
+            continue;
```

Comment:
When do we encounter these negative indices?

Reply:
You may see the negative indexes when it is batched sparse matrices and number of non-zero elements in each sparse matrix differ.